Control flow graph for uninitialized storage return detection.

ekpyron opened the pull request:
Closes #2628.

ekpyron committed:
Add control flow graph.

ekpyron committed:
Add control flow analyzer and test for uninitialized storage returns.

ekpyron committed:
Add Changelog entry.

ekpyron commented:
1b25a528d3e4e6c86a38586ed07f4244f771a829 and 256f00f09afd9b60e2e16916dc4b9790dc661d66 could be squashed before merging. (The latter basically refactors the former.)

I could also split this PR into adding the control flow graph and using it for detecting uninitialized storage returns (at the moment it's already separate commits and separate files).

chriseth commented on libsolidity/interface/CompilerStack.cpp:
```diff
@@ -222,6 +224,22 @@ bool CompilerStack::analyze()
+			{
+				ControlFlowAnalyzer controlFlowAnalyzer(cfg, m_errorReporter);
+				for (Source const* source: m_sourceOrder)
+					if (!controlFlowAnalyzer.analyze(*source->ast))
```
Might be unnecessary optimization, but if the analyzer does not care about cross-source control-flow, we could change this into a single loop, where we first construct the CFG for a single source file and (if that worked), run the analyzer on it.

ekpyron replied:
Modifiers may be defined in a base contract in a different source file, so we cannot construct/analyze control flow for each source file separately.

chriseth commented on libsolidity/analysis/ControlFlowGraph.h:
```diff
@@ -0,0 +1,140 @@
+namespace solidity
+{
+
+/// Basic Control Flow Block.
```
Usually we use `/** ... */` to document classes and structs and `///` to document functions.

chriseth commented on libsolidity/analysis/ControlFlowGraph.h:
```diff
@@ -0,0 +1,140 @@
+
+/// Basic Control Flow Block.
+/// Basic block of control flow. Consists of a set of AST nodes
+/// for which control flow is always linear.
```
Perhaps clarify that `revert` is seen as breaking control flow, this is not always the case in other modules.

Also: How do you handle function calls?

ekpyron replied:
Function calls are ignored at the moment (i.e. not treated as breaking control flow). The only exceptions are revert, assert and require.

chriseth commented on libsolidity/analysis/ControlFlowGraph.h:
```diff
@@ -0,0 +1,140 @@
+/// for which control flow is always linear.
+struct ControlFlowBlock
+{
+	/// All variable declarations executed in this control flow block.
```
What does `executed` mean? Is this about the first assignment? How does this handle the new and old scoping rules?

ekpyron replied:
``executed`` just means ``present inside the block``. This is not used for anything yet - I just added it already, since it might be a use case in the future (and with new scoping rules). Also note that control flow blocks are always completely inside one scope (I also documented that).

chriseth commented on libsolidity/analysis/ControlFlowGraph.h:
```diff
@@ -0,0 +1,140 @@
+{
+	/// All variable declarations executed in this control flow block.
+	std::vector<VariableDeclaration const*> variableDeclarations;
+	/// All expressions executed in this control flow block (this includes all subexpressions!).
```
`evaluated` instead of `executed`?

chriseth replied:
What is the order of the vector?

ekpyron replied:
I could specify the current order of these vectors, but IMO order does not and should not matter for a control flow graph. It's not an AST - the only question the CFG wants to answer is whether something is present in a path of control flow or not. I documented them as "unordered" now.

chriseth commented on libsolidity/analysis/ControlFlowGraph.h:
```diff
@@ -0,0 +1,140 @@
+	/// All expressions executed in this control flow block (this includes all subexpressions!).
+	std::vector<Expression const*> expressions;
+	/// All inline assembly statements executed in this control flow block.
+	std::vector<InlineAssembly const*> inlineAssemblyStatements;
```
Again: order?

chriseth commented on libsolidity/analysis/ControlFlowGraph.h:
```diff
@@ -0,0 +1,140 @@
+struct CFGNode
+{
+	/// Entry nodes. All CFG nodes from which control flow may move into this node.
+	std::vector<CFGNode*> entries;
```
Does it make sense to just add these as members to ControlFlowBlock?

ekpyron replied:
I intentionally divided the node, i.e. the graph structure, from its contents. This may allow further optimizations in the future. For example at the moment the modifier flow is copied entirely when "applying" the modifier to the function flow - in the future only the graph structure could be duplicated, but the contained ControlFlowBlock could be the same object.

chriseth commented on libsolidity/analysis/ControlFlowGraph.h:
```diff
@@ -0,0 +1,140 @@
+		entry(_entry), exit(_exit), exception(_exception) {}
+	/// Entry node. Control flow of the function starts here.
+	/// This node does not have any entries.
+	CFGNode* entry;
```
This might be confusing when compared with CFGNode. Here, it designates the first node of the function, while in CFGNode it designates the predecessors.

ekpyron replied:
In both cases it's the node from which control flow enters - so I would say it's exactly the same as in CFGNode and hence it makes sense to use the same name.

chriseth commented on libsolidity/analysis/ControlFlowGraph.h:
```diff
@@ -0,0 +1,140 @@
+	CFGNode* entry;
+	/// Exit node. Control flow of the function ends here.
+	/// This node does not have any exits, but may have multiple entries
+	/// (e.g. all return statements of the function).
```
Perhaps it would be good to be even more explicit here: `The end of the function's body and all return statements, revert, throw, assert, etc are connected to this node.`  - do reverts end up here? If not, the above line should probably read "Non-reverting control flow of the function ends here."

chriseth replied:
Perhaps for both entry and exit, it might be good to state that they are empty nodes.

chriseth commented on libsolidity/analysis/ControlFlowGraph.h:
```diff
@@ -0,0 +1,140 @@
+	CFGNode* exit;
+	/// Exception node. Control flow of the function in case of revert.
+	/// This node does not have any exists, but may have multiple entries
+	/// (e.g. all revert, assert and require statements).
```
Also throw?

chriseth commented on libsolidity/analysis/ControlFlowGraph.h:
```diff
@@ -0,0 +1,140 @@
+	/// Exception node. Control flow of the function in case of revert.
+	/// This node does not have any exists, but may have multiple entries
+	/// (e.g. all revert, assert and require statements).
+	CFGNode* exception;
```
Perhaps name this revert - exceptions might have a different meaning in the future.

chriseth commented on libsolidity/analysis/ControlFlowGraph.h:
```diff
@@ -0,0 +1,140 @@
+};
+
+/// Describes the control flow of a modifier.
+/// Inherits all members from FunctionFlow.
```
This is redundant.

chriseth commented on libsolidity/analysis/ControlFlowGraph.h:
```diff
@@ -0,0 +1,140 @@
+
+/// Describes the control flow of a modifier.
+/// Inherits all members from FunctionFlow.
+struct ModifierFlow : FunctionFlow
```
No space in front of `:`

chriseth commented on libsolidity/analysis/ControlFlowGraph.h:
```diff
@@ -0,0 +1,140 @@
+	template<typename... Args>
+	ModifierFlow(Args&&... args): FunctionFlow(std::forward<Args>(args)...) {}
+
+	CFGNode* placeholderEntry = nullptr;
```
Please document, again, it is not 100% clear which perspective we use here.

chriseth commented on libsolidity/analysis/ControlFlowGraph.h:
```diff
@@ -0,0 +1,140 @@
+	CFGNode* placeholderExit = nullptr;
+};
+
+class CFG: private ASTConstVisitor
```
Perhaps `CFGConstructor`?

chriseth commented on libsolidity/analysis/ControlFlowGraph.h:
```diff
@@ -0,0 +1,140 @@
+
+	ErrorReporter& m_errorReporter;
+
+	/// Maps function definitions to their control flow.
```
Redundant.

chriseth commented on libsolidity/analysis/ControlFlowGraph.h:
```diff
@@ -0,0 +1,140 @@
+
+	/// Maps function definitions to their control flow.
+	std::map<FunctionDefinition const*, std::shared_ptr<FunctionFlow>> m_functionControlFlow;
+	/// Maps modifier definitions to their control flow.
```
Redundant

chriseth commented on libsolidity/analysis/ControlFlowGraph.h:
```diff
@@ -0,0 +1,140 @@
+{
+	FunctionFlow(CFGNode* _entry, CFGNode* _exit, CFGNode* _exception):
+		entry(_entry), exit(_exit), exception(_exception) {}
+	/// Entry node. Control flow of the function starts here.
```
virtual destructor

chriseth reviewed (changes requested):
Not yet reviewed everything.

ekpyron commented on libsolidity/analysis/ControlFlowGraph.h:
```diff
@@ -0,0 +1,150 @@
+	/// The control flow of the modifier that is currently parsed.
+	std::shared_ptr<ModifierFlow> m_currentModifierFlow;
+
+	friend class ControlFlowParser;
```
Usually I would not use ``friend`` declarations either, but in this case since ``ControlFlowParser`` is a class that is local to the implementation I would make an exception. Moving the entire declaration of ``ControlFlowParser`` to an inline class or creating a public interface in ``CFG`` for accessing e.g. ``m_currentFunctionFlow`` seems messy in comparison.

If you disagree I can refactor the friend class away, though :-).

chriseth commented on libsolidity/analysis/ControlFlowGraph.h:
```diff
@@ -0,0 +1,150 @@
+	virtual ~FunctionFlow() {}
+	/// Entry node. Control flow of the function starts here.
+	/// This node is empty and does not have any entries.
+	CFGNode* entry;
```
Please use ` = nullptr;`

chriseth commented on libsolidity/analysis/ControlFlowGraph.h:
```diff
@@ -0,0 +1,150 @@
+	/// Exit node. All non-reverting control flow of the function ends here.
+	/// This node is empty and does not have any exits, but may have multiple entries
+	/// (e.g. all return statements of the function).
+	CFGNode* exit;
```
Please use ` = nullptr;`

chriseth commented on libsolidity/analysis/ControlFlowGraph.h:
```diff
@@ -0,0 +1,150 @@
+	/// Revert node. Control flow of the function in case of revert.
+	/// This node is empty does not have any exists, but may have multiple entries
+	/// (e.g. all assert, require, revert and throw statements).
+	CFGNode* revert;
```
Please use ` = nullptr;`

chriseth commented on libsolidity/analysis/ControlFlowGraph.h:
```diff
@@ -0,0 +1,150 @@
+	/// (e.g. all return statements of the function).
+	CFGNode* exit;
+	/// Revert node. Control flow of the function in case of revert.
+	/// This node is empty does not have any exists, but may have multiple entries
```
Typo: exists

chriseth commented on libsolidity/analysis/ControlFlowGraph.h:
```diff
@@ -0,0 +1,150 @@
+private:
+	CFGNode* newNode();
+
+	/// Initially the control flow for all functions *without* modifiers and for
```
Perhaps replace `without` by `ignoring` to reduce ambiguity.

chriseth commented on libsolidity/analysis/ControlFlowGraph.cpp:
```diff
@@ -0,0 +1,552 @@
+{
+	_astRoot.accept(*this);
+	applyModifiers();
+	return Error::containsOnlyWarnings(m_errorReporter.errors());
```
Strictly, this should only return false if it added new errors.

ekpyron replied:
I could change that here or I could break on the first error in ``CompilerStack.cpp`` - however, the current behaviour is consistent with the other analysis modules, see e.g. ``PostTypeChecker``, so we should probably either leave it as it is here or change it everywhere...

chriseth replied:
Yeah, it's probably fine.

chriseth commented on libsolidity/analysis/ControlFlowGraph.cpp:
```diff
@@ -0,0 +1,552 @@
+
+bool CFG::visit(ModifierDefinition const& _modifier)
+{
+	m_currentModifierFlow = make_shared<ModifierFlow>(newNode(), newNode(), newNode());
```
Please assert that the pointers are empty.

chriseth commented on libsolidity/analysis/ControlFlowGraph.cpp:
```diff
@@ -0,0 +1,552 @@
+
+bool CFG::visit(FunctionDefinition const& _function)
+{
+	m_currentFunctionFlow = make_shared<FunctionFlow>(newNode(), newNode(), newNode());
```
Please assert that the pointers are empty.

chriseth commented on libsolidity/analysis/ControlFlowGraph.cpp:
```diff
@@ -0,0 +1,552 @@
+
+	while (!nodesToClone.empty())
+	{
+		auto srcNode = nodesToClone.top();
```
I'm sorry, but I am really not fond of `auto`, I constantly have to look up the type to check whether it performs a copy or not.

chriseth commented on libsolidity/analysis/ControlFlowGraph.cpp:
```diff
@@ -0,0 +1,552 @@
+
+		solAssert(oldToNew.count(srcNode), "");
+
+		auto dstNode = oldToNew[srcNode];
```
Shouldn't that be called `newNode` (and the other one `oldNode`)? The current naming suggests that control flow goes from `srcNode` to `dstNode`.

chriseth commented on libsolidity/analysis/ControlFlowGraph.cpp:
```diff
@@ -0,0 +1,552 @@
+
+		auto dstNode = oldToNew[srcNode];
+
+		dstNode->block = srcNode->block;
```
Ok, here it makes a lot of sense to have `block` being a member :)

chriseth commented on libsolidity/analysis/ControlFlowGraph.cpp:
```diff
@@ -0,0 +1,552 @@
+		auto dstNode = oldToNew[srcNode];
+
+		dstNode->block = srcNode->block;
+		for (auto& entry: srcNode->entries)
```
Why not `const&`?

chriseth commented on libsolidity/analysis/ControlFlowGraph.cpp:
```diff
@@ -0,0 +1,552 @@
+	if (!oldToNew.count(_modifierFlow.exit))
+		oldToNew[_modifierFlow.exit] = newNode();
+
+	_functionFlow->entry = oldToNew[_modifierFlow.entry];
```
Do the old nodes that were cloned from still exist? They are fully disconnected and could be deleted, right?

ekpyron replied:
The old nodes (i.e. the modifier nodes) aren't fully disconnected - they are unchanged.

The idea for applying a modifier to a function is to create a copy of the modifier flow and - while copying - to replace the placeholder entry and exit with the function's entry and exit (and also replace the revert node of the modifer with the revert node of the function).

I still need the modifier flow afterwards, since the same modifier may be applied to several functions. I'd also keep it after all modifiers have been applied, because we may want to analyze modifier flow independently of function flow as well at some point.

I will try to clarify what happens by renaming the local variables here. ``oldToNew`` should probably rather be called something like ``copySrcToCopyDst`` and I'll probably not rename ``srcNode`` to ``oldNode`` and ``dstNode`` to ``newNode``, but instead ``srcNode`` to ``copySrcNode`` and ``dstNode`` to ``copyDstNode``. Do you think that would clarify what happens?

ekpyron replied:
I'll also add some documentation in the header.

chriseth replied:
I agree about the modifier nodes, but what about the function nodes?

chriseth reviewed (changes requested):
Another partial review.

chriseth commented on libsolidity/analysis/ControlFlowGraph.cpp:
```diff
@@ -0,0 +1,552 @@
+	/// m_currentNode is set to nullptr and has to be set manually or
+	/// using mergeFlow later.
+	template<size_t n>
+	std::array<CFGNode*, n> splitFlow()
```
Does this need to be a template?

chriseth replied:
Ok, it does not hurt because it is a cpp file anyway.

ekpyron replied:
Yeah. The number of paths to split into is always known at compile time and this function is only used locally in this cpp file, so I would say a template makes sense. However, it turns out that so far it's actually only used with ``n = 2`` in the end, so I could also just fix that and make it return a ``pair`` instead. I thought it made sense to keep it general, though.

chriseth replied:
It's fine!

chriseth commented on libsolidity/analysis/ControlFlowGraph.cpp:
```diff
@@ -0,0 +1,552 @@
+
+	if (_ifStatement.falseStatement())
+	{
+		auto nodes = splitFlow<2>();
```
Code duplication.

chriseth commented on libsolidity/analysis/ControlFlowGraph.cpp:
```diff
@@ -0,0 +1,552 @@
+{
+	solAssert(!!m_currentNode, "");
+
+	if (auto initializationExpression = _forStatement.initializationExpression())
```
If we ever change the return type here, this might perform a copy without us noticing. I think it's fair to just call the function twice...

ekpyron replied:
If the return type is changed to something that might perform a copy, the pointer dereference in the next line should fail, though. I can change to ``auto const*``, unless that's still too much ``auto`` :)..

chriseth commented on libsolidity/analysis/ControlFlowGraph.cpp:
```diff
@@ -0,0 +1,552 @@
+
+	auto condition = createLabelHere();
+
+	if (auto conditionExpression = _forStatement.condition())
```
Also here.

chriseth commented on libsolidity/analysis/ControlFlowGraph.cpp:
```diff
@@ -0,0 +1,552 @@
+		auto whileBody = createLabelHere();
+
+		{
+			BreakContinueScope scope(*this, afterWhile, whileBody);
```
Actually - does `continue` really go to `whileBody` and not to `condition`? Might that be a bug in the code generator?

ekpyron replied:
This is a ``do { ... } while(...);``. For those ``continue`` goes back to the body and that's what I would expect, so I don't think it's a bug. In the ``while(...) { ... }`` case it does indeed go to ``whileCondition`` instead.

chriseth replied:
Turns out, in C and Javascript the `continue` statement jumps to right before the condition, while in Solidity, it jumps to the beginning of the loop body. We should change that with 0.5.0.

chriseth commented on libsolidity/analysis/ControlFlowGraph.cpp:
```diff
@@ -0,0 +1,552 @@
+
+		auto nodes = splitFlow<2>();
+
+		auto& whileBody = nodes[0];
```
Why do they need to be references?

chriseth commented on libsolidity/analysis/ControlFlowGraph.cpp:
```diff
@@ -0,0 +1,552 @@
+	solAssert(!!m_currentNode, "");
+	solAssert(!!m_cfg.m_currentFunctionFlow, "");
+	solAssert(!!m_cfg.m_currentFunctionFlow->exit, "");
+	// only the first return statement is interesting
```
I do not understand this comment.

chriseth replied:
Shouldn't there be an assert if there are multiple?

ekpyron replied:
That's true - two consecutive ``return``'s in the same scope used to end up in the same block, but I later changed that, so that after the first return a new disconnected block is started, in which the second return ends up, which makes much more sense, but I forgot to remove the comment, sorry about that.

chriseth commented on libsolidity/analysis/ControlFlowGraph.cpp:
```diff
@@ -0,0 +1,552 @@
+	solAssert(!!m_currentNode, "");
+	if (auto const* expression = dynamic_cast<Expression const*>(&node))
+		m_currentNode->block.expressions.emplace_back(expression);
+	if (auto const* variableDeclaration = dynamic_cast<VariableDeclaration const*>(&node))
```
Please use `else if` here.

chriseth commented on libsolidity/analysis/ControlFlowGraph.cpp:
```diff
@@ -0,0 +1,552 @@
+		m_currentNode->block.expressions.emplace_back(expression);
+	if (auto const* variableDeclaration = dynamic_cast<VariableDeclaration const*>(&node))
+		m_currentNode->block.variableDeclarations.emplace_back(variableDeclaration);
+	if (auto const* assembly = dynamic_cast<InlineAssembly const*>(&node))
```
Also here.

chriseth commented on libsolidity/analysis/ControlFlowGraph.cpp:
```diff
@@ -0,0 +1,552 @@
+bool ControlFlowParser::visit(FunctionCall const& _functionCall)
+{
+	solAssert(!!m_currentNode, "");
+	if (auto functionType = dynamic_pointer_cast<FunctionType const>(_functionCall.expression().annotation().type))
```
Please add an assert that `annotation().type` is not null.

chriseth commented on libsolidity/analysis/ControlFlowGraph.cpp:
```diff
@@ -0,0 +1,552 @@
+	if (auto functionType = dynamic_pointer_cast<FunctionType const>(_functionCall.expression().annotation().type))
+		switch (functionType->kind())
+		{
+			case FunctionType::Kind::Revert:
```
Can you still visit the expression in case of revert, require and assert? We might support weird stuff in the future, just to make sure.

chriseth commented on libsolidity/analysis/ControlFlowGraph.h:
```diff
@@ -0,0 +1,150 @@
+
+	/// The control flow of the function that is currently parsed.
+	/// Note: this can also be a ModifierFlow
+	std::shared_ptr<FunctionFlow> m_currentFunctionFlow;
```
Why does the CFG own these two variables in the first place? Wouldn't it be better to move them to the parser?

ekpyron replied:
That's true, I'll first take care of everything else and then think about this. In any case, unfortunately, I still think it makes sense for ``CFG`` to at least own all nodes and that ``CFG::newNode`` shouldn't be public, so we'd still need the ``friend`` declaration.

chriseth replied:
If these are the only ones, then it might make sense to create a separate class `NodeContainer` that owns the nodes and has a `newNode` function and is a member of CFG, but is passed on to the parser via reference.

chriseth replied:
(private member)

ekpyron replied:
Yes, right, this might be the cleanest solution. I'm working on it.

chriseth commented on libsolidity/analysis/ControlFlowAnalyzer.cpp:
```diff
@@ -0,0 +1,129 @@
+{
+	auto const& functionFlow = m_cfg.functionFlow(_function);
+	checkUnassignedStorageReturnValues(_function, functionFlow.entry, functionFlow.exit);
+	return ASTConstVisitor::visit(_function);
```
This will also visit all sub nodes. Is this necessary?

chriseth commented on libsolidity/analysis/ControlFlowAnalyzer.cpp:
```diff
@@ -0,0 +1,129 @@
+	{
+		auto& unassignedAtFunctionEntry = unassigned[_functionEntry];
+		for (auto const& returnParameter: _function.returnParameterList()->parameters())
+			if (returnParameter->type()->dataStoredIn(DataLocation::Storage))
```
This is the place where we could satisfy the other feature request. The question is whether we want it or not :)

chriseth replied:
Is there a test that uses the default location?

ekpyron replied:
``DataLocation`` in contrast to ``VariableDeclaration::Location`` does not have a default - the default should already be resolved here. I'll add tests using the default location regardless.

And yes - there's a lot of room for further checks in future PRs here :).

chriseth commented on libsolidity/analysis/ControlFlowAnalyzer.cpp:
```diff
@@ -0,0 +1,129 @@
+		}
+	}
+
+	if (unassigned[_functionExit].size() > 0)
```
`!...empty()`?

chriseth commented on libsolidity/analysis/ControlFlowAnalyzer.cpp:
```diff
@@ -0,0 +1,129 @@
+			}
+		);
+		for (auto returnVal: unassignedOrdered)
+			m_errorReporter.warning(returnVal->location(), "uninitialized storage pointer may be returned");
```
Error messages should be full sentences starting with a capitalized word and ending in a period :)

chriseth replied:
Do we want to make this an error for 0.5.0?

chriseth replied:
Also would it be possible to add a secondary source location pointing to the place where control flow returns with this variable unassigned?

chriseth replied:
Ah and another suggestion for the message: `This variable is of storage pointer type and might be returned uninitialized.`

ekpyron replied:
I don't think we can make this an error for 0.5.0 yet. There are still corner cases where the current analysis will result in false positives, e.g.:

```
contract C {
  function alwaysRevertWithSomeSpecificReason() {
    revert(...);
  }
  struct S { uint a; };
  S s;
  function f(bool flag) returns (S storage r) {
    if (flag) r = s;
    else alwaysRevertWithSomeSpecificReason();
  }
}
```
This example would require cross-function control flow, but even if we had that, I don't think it should be an error - consider:
```
contract C {
  struct S { uint a; };
  S s1;
  S s2;
  function f(bool a, bool b) returns (S storage r) {
    require(a || b);
    if (a) r = s1;
    if (b) r = s2;
  }
}
```
Using the SMT solver may help for such cases in the future (the CFG is rather dumb without it), but I'm not sure whether we can ever catch all false positives and at least for now I don't think it can be an error.

chriseth replied:
We have the policy that every warning can be silenced by a simple code change. The obvious change here would be to use `r = r` in problematic code paths. This is even a stack assignment, so it has no practical influence on gas costs. So I think the false positives should not be an issue here.

What could be an argument is that everything that is an error has to be specified somehow, and it is probably very hard to formalize the control flow graph analysis.

chriseth commented on libsolidity/analysis/ControlFlowAnalyzer.cpp:
```diff
@@ -0,0 +1,129 @@
+			// kill all return values to which a value is assigned
+			for (auto expression: node->block.expressions)
+			{
+				if (auto const* assignment = dynamic_cast<Assignment const*>(expression))
```
Does it make sense to extract some of this into a function?
You could do `unasignedAtNode -= assignedInExpression()` if we have an overload there :)

chriseth commented on libsolidity/analysis/ControlFlowAnalyzer.cpp:
```diff
@@ -0,0 +1,129 @@
+			auto& unassignedAtExit = unassigned[exit];
+			auto oldSize = unassignedAtExit.size();
+			unassignedAtExit.insert(unassignedAtNode.begin(), unassignedAtNode.end());
+			// traverse an exit, if we are on a path with new unassigned return values to consider
```
Should this be "re-traverse"?

ekpyron replied:
"(re)traverse" :) - it can either be the first traversal or traversing it again. During the first traversal of an exit, ``unassignedAtExit`` will be empty.

ekpyron commented:
@chriseth I just pushed a commit that moves m_currentFunctionFlow and refactors CFG and ControlFlowParser around that change (I think it's indeed much cleaner now).
I also used references instead of pointers and unique pointers instead of shared pointers, which clarifies ownership and lifetime further. I think (resp. hope :)) this is the last larger change we need here.

Regarding the secondary source locations:
Yes, I wasn't sure about the "end of function" location either, I'm not sure what's the best alternative, though (the last statement before the end of the function is not really the right location either). I could instead only give secondary source locations for return statements and not for the end of the function...

``might be returned without assignment`` may be even clearer, that's true, I'll change it.

chriseth commented:
Yes, perhaps just modify the message to contain `at the end of the function` without giving a secondary source location.

chriseth commented:
I'm just wondering, is the warning message clear enough? Do people realize what happens if they return a storage pointer variable without assigning it?

ekpyron commented:
Changing the error message to include "at the end of the function" won't work - it may be the case that both at a return statement *and* at the end of the function a storage pointer is left unassigned.
I could use the whole function as secondary source location for the end-of-function-case, though.

chriseth commented on libsolidity/analysis/ControlFlowGraph.cpp:
```diff
@@ -0,0 +1,566 @@
+{
+
+/** Helper class that parses the control flow of a non-top-level ASTNode. */
+class ControlFlowParser: private ASTConstVisitor
```
Since it is not a friend anymore, it might make sense to move it to its own file. Also perhaps rather call it `Builder` than `Parser`?

ekpyron replied:
OK, I can do that :).

ekpyron commented on libsolidity/analysis/ControlFlowBuilder.h:
```diff
@@ -0,0 +1,140 @@
+	/// m_currentNode is set to nullptr and has to be set manually or
+	/// using mergeFlow later.
+	template<size_t n>
+	std::array<CFGNode*, n> splitFlow()
```
Well... now these templates are in a header :).

ekpyron commented:
In the last commit I extracted ``ControlFlowParser`` into a separate file and renamed it to ``ControlFlowBuilder``. However, I did that largely automatically and haven't yet rechecked, that it didn't mess up code formatting at some places, but I have to stop for today.

ekpyron committed:
Add test for default location.

ekpyron commented:
Rebased and ready for review.

chriseth commented on libsolidity/analysis/ControlFlowGraph.h:
```diff
@@ -0,0 +1,147 @@
+
+	FunctionFlow const& functionFlow(FunctionDefinition const& _function) const;
+
+	class NodeContainer {
```
Newline before `{`

chriseth commented on libsolidity/analysis/ControlFlowBuilder.cpp:
```diff
@@ -0,0 +1,370 @@
+using namespace solidity;
+using namespace std;
+
+ControlFlowBuilder::ControlFlowBuilder(CFG::NodeContainer& _nodeContainer, FunctionFlow const& _functionFlow) :
```
No space in front of `:`

chriseth commented on libsolidity/analysis/ControlFlowAnalyzer.cpp:
```diff
@@ -0,0 +1,154 @@
+
+			m_errorReporter.warning(
+				returnVal->location(),
+				"This variable is of storage pointer type and might be returned without assignment.",
```
What about `This variable is of storage pointer type and might be returned without assignment. This can cause storage corruption. Assign the variable (potentially from itself) to remove this warning.`

chriseth reviewed (dismissed):
Final tweaks.
